### PR TITLE
[Yaml] dump non UTF-8 encoded strings as binary data

### PR DESCRIPTION
**Pull request by xabbuh**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | #18241 |
| License | MIT |
| Doc PR |  |

### CODE REVIEW

File: src/Symfony/Component/Yaml/Inline.php [hunk 1]
```diff
@@ -203,6 +203,13 @@ public static function dump($value, $flags = 0)
                 return "''";
             case Yaml::DUMP_BASE64_BINARY_DATA & $flags && self::isBinaryString($value):
                 return '!!binary '.base64_encode($value);
+        }
+
+        if (self::isBinaryString($value)) {
+            @trigger_error('Dumping non UTF-8 data without passing the DUMP_BASE64_BINARY_DATA flag is deprecated since Symfony 3.1 and will be removed in 4.0.', E_USER_DEPRECATED);
```
Comment (fabpot) [inline src/Symfony/Component/Yaml/Inline.php]: I'm wondering if `Yaml::DUMP_BASE64_BINARY_DATA` should not be set by default in 4.0 and the option deprecated. that way, the YAML dumper always behaves as expected without having to change some options which seems weird.

Comment (xabbuh) [inline src/Symfony/Component/Yaml/Inline.php]: I think that's a good idea. I updated the deprecation message to reflect that.

**Comment by fabpot** [comment]
Can we also deprecate `DUMP_BASE64_BINARY_DATA` here as well?

**Comment by xabbuh** [comment]
Imo that does not make much sense. We will introduce the flag in 3.1 and people will need it to be able to dump binary. Deprecating it in the same instance would be weird for users of the dumper. And we cannot dump them as binary strings by default as that would be a BC break, wouldn't it?

**Comment by fabpot** [comment]
I forgot that this was introduced in 3.1. So, I propose to drop it altogether and always do the right thing by default. As Symfony YAML load is able to deal with binaries, I don't see any BC break.

**Comment by xabbuh** [comment]
Fair enough. I removed the flag, fixed the tests, and updated the changelog.

**Comment by fabpot** [comment]
:+1:

File: src/Symfony/Component/Yaml/Inline.php [hunk 2]
```diff
@@ -627,7 +627,7 @@ public static function evaluateBinaryScalar($scalar)
 
     private static function isBinaryString($value)
     {
-        return preg_match('/[^\x09-\x0d\x20-\xff]/', $value);
+        return !preg_match('//u', $value);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Inline.php]: maybe we should include also strings that contains control chars (with a few excluded line CR, LF, TAB)?

Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Inline.php]: In similar cases, I usually use:
`return !preg_match('//u', $value) || preg_match('/[\x00-\x08\x0B\x0E-\x1A\x1C-\x1F\x7F]+/', $value);`

Comment (xabbuh) [inline src/Symfony/Component/Yaml/Inline.php]:
```php
return !preg_match('//u', $value) || preg_match('/[^\x09-\x0d\x20-\xff]/', $value);
```

Do have something like this in mind?

Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Inline.php]: the above regexp is better I think (not invented here :) )

Comment (xabbuh) [inline src/Symfony/Component/Yaml/Inline.php]: Oh sorry, I missed your second comment. The previous regex was borrowed from PHPUnit. But if you have good experience with the other one, I am fine with using it.

Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Inline.php]: :+1: for phpunit's one

**Comment by xabbuh** [comment]
@nicolas-grekas I updated the check.

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Comment by alexpott** [comment]
This breaks a feature in Drupal 8's configuration management that relies on Symfony's YAML to dump configuration. One of the main advantages of YAML is that it is human readable. We were encoding multiple plural forms like so:
`format_plural_string: "1 place\x03@count places"`
This becomes:
`format_plural_string: !!binary MSBwbGFjZQNAY291bnQgcGxhY2Vz`

We've hit this whilst trying to move Drupal 8 to Symfony 3.2

**Comment by stof** [comment]
do you really rely on non-unicode chars to split the plurals ?

**Comment by xabbuh** [comment]
@alexpott Thanks for reporting this. The thing is that `\x03` is no valid unicode character. So the fact that this somehow worked before was more by accident (http://yaml-online-parser.appspot.com/, for example, will give you an error like `unacceptable character #x0003: special characters are not allowed in "<unicode string>", position 30`).

**Pull request merged**